### PR TITLE
fix: UI/UX fixes #7156

### LINKED ISSUE #7156: 🐛 Bug: UI/UX fixes
### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Steps to reproduce ?

- [x] `Group-by` initial menu to be made similar to `Sort`
Screenshot 2023-12-05 at 2 40 12 PM
Screenshot 2023-12-05 at 2 40 24 PM

- [x] Update `Group-by` testcases

- [x] DateTime field, on the picker - set default time to current time always & enable `OK` button by default

- [x] Default name for token, let it be `Token-1` `Token-2` `Token-3` and such
Screenshot 2023-12-06 at 12 35 13 PM

- [x] Webhook modal
  - [x] Default name, let it be `Webhook-1` `Webhook-2` and such
  - [x] Set first option in the trigger list as selected by default
Screenshot 2023-12-06 at 12 37 52 PM
- [x] Created-at, Updated-at : right click should disable `Clear` option;
- [x] Introduce paste option in cell right click context menu
- [x] System column shouldn't be editable
- [x] Junction table shouldn't be editable
Screenshot 2023-12-06 at 4 06 35 PM
Screenshot 2023-12-06 at 4 06 51 PM

- [x] Minimum width for select drop down's
Screenshot 2023-12-07 at 9 57 09 PM

### Desired Behavior

x

### Project Details

202.9

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by rameshmane7218**
## Change Summary

- UI/UX bug fixes #7156 

## Change type

- [x] fix: (bug fix for the user, not a fix to a build script)
- [x] fix: `rolePermissions` in `lib/types.ts` file is imported from wrong file `./constants` insted of `./acl`
- [x] fix: Audit log for external database not visible through UI #7165

## Test/ Verification

Provide summary of changes.

### CODE REVIEW

File: packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+const onArrowUp = () => {
+  activeFieldIndex.value = Math.max(activeFieldIndex.value - 1, 0)
+}
+</script>
+
+<template>
+  <div
+    class="flex flex-col w-full pt-4 pb-2 min-w-64 nc-group-create-modal"
```
Comment (wingkwong) [inline packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue:86]: `nc-group-by-create-modal`

File: packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    <div class="flex pb-3 px-4 border-b-1 border-gray-100">
+      <!-- Todo language for selectFieldToGroup -->
+      <input ref="inputRef" v-model="search" class="w-full focus:outline-none" :placeholder="$t('msg.selectFieldToGroup')" />
+    </div>
+    <div class="flex-col w-full max-h-100 max-w-76 nc-scrollbar-md !overflow-y-auto">
+      <div v-if="!options.length" class="flex text-gray-500 px-4 py-2.25">{{ $t('general.empty') }}</div>
+      <div
+        v-for="(option, index) in options"
+        :key="index"
+        v-e="['c:sort:add:column:select']"
```
Comment (wingkwong) [inline packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue:100]: name another name

File: packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        class="flex flex-row h-10 items-center gap-x-1.5 px-2.5 hover:bg-gray-100 cursor-pointer nc-group-column-search-item"
```
Comment (wingkwong) [inline packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue:101]: group -> group-by

File: packages/nc-gui/components/smartsheet/toolbar/GroupByMenu.vue [hunk 4]
```diff
@@ -157,27 +180,29 @@ const loadAllowedLookups = async () => {
     for (const col of meta.value?.columns || []) {
       if (col.uidt !== UITypes.Lookup) continue
 
-      let nextCol = col
+      let nextCol: ColumnType | undefined = col
```
Comment (wingkwong) [inline packages/nc-gui/components/smartsheet/toolbar/GroupByMenu.vue:183]: in what case nextCol can be `undefined`?

File: packages/nc-gui/components/smartsheet/toolbar/GroupByMenu.vue [hunk 5]
```diff
@@ -265,23 +296,31 @@ watch(meta, async () => {
...
+          v-model:visible="showCreateGroupBy"
+          :trigger="['click']"
+          overlay-class-name="nc-toolbar-dropdown"
         >
-          <div class="flex gap-1 items-center" :class="{ 'text-brand-500': groupedByColumnIds.length >= _groupBy.length }">
-            <div class="flex">
-              {{ $t('activity.addSubGroup') }}
+          <NcButton
+            v-e="['c:group-by:add']"
+            class="nc-add-group-btn !text-brand-500"
```
Comment (wingkwong) [inline packages/nc-gui/components/smartsheet/toolbar/GroupByMenu.vue:307]: group-by

File: packages/nc-gui/components/smartsheet/header/Cell.vue [hunk 6]
```diff
@@ -48,7 +49,7 @@ const closeAddColumnDropdown = () => {
 }
 
 const openHeaderMenu = () => {
-  if (isLocked.value) return
+  if (isLocked.value || (!!isSystemColumn(column.value) && column.value.uidt !== UITypes.ID)) return
```
[the remark this answers is not in the document]
Comment (rameshmane7218) [reply packages/nc-gui/components/smartsheet/header/Cell.vue]: fixed

File: packages/nc-gui/components/webhook/Editor.vue [hunk 7]
```diff
@@ -472,6 +474,30 @@ async function testWebhook() {
   await webhookTestRef.value.testWebhook()
 }
 
+function getDefaultTokenValue(hooks: HookType[]) {
```
[the remark this answers is not in the document]
Comment (rameshmane7218) [reply packages/nc-gui/components/webhook/Editor.vue]: Created `extractNextDefaultName` generic function and added in helper parsers

File: packages/nc-gui/components/account/Token.vue [hunk 8]
```diff
@@ -42,6 +42,30 @@ const pagination = reactive({
...
+        }
+        return 0
+      }) || []
+
+  if (extractedSortedTokenNumbers.length) {
+    return `${defaultTokenName}-${(extractedSortedTokenNumbers[extractedSortedTokenNumbers.length - 1] || 0) + 1}`
+  } else {
+    return `${defaultTokenName}-1`
+  }
+}
```
[the remark this answers is not in the document]
Comment (rameshmane7218) [reply packages/nc-gui/components/account/Token.vue]: done

File: packages/nc-gui/components/smartsheet/grid/Table.vue [hunk 9]
```diff
@@ -1703,6 +1703,7 @@ onKeyStroke('ArrowDown', onDown)
                 (isLinksOrLTAR(fields[contextMenuTarget.col]) || !isVirtualCol(fields[contextMenuTarget.col]))
               "
               class="nc-base-menu-item"
+              :disabled="!!isSystemColumn(fields[contextMenuTarget.col])"
```
[the remark this answers is not in the document]
Comment (rameshmane7218) [reply packages/nc-gui/components/smartsheet/grid/Table.vue]: done

File: packages/nc-gui/components/smartsheet/grid/Table.vue [hunk 10]
```diff
@@ -1716,6 +1717,7 @@ onKeyStroke('ArrowDown', onDown)
               v-else-if="contextMenuTarget && hasEditPermission"
               class="nc-base-menu-item"
               @click="clearSelectedRangeOfCells()"
+              :disabled="!!isSystemColumn(fields[contextMenuTarget.col])"
```
[the remark this answers is not in the document]
Comment (rameshmane7218) [reply packages/nc-gui/components/smartsheet/grid/Table.vue]: done

File: packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+<template>
+  <div
+    class="flex flex-col w-full pt-4 pb-2 min-w-64 nc-group-create-modal"
+    @keydown.arrow-down.prevent="onArrowDown"
+    @keydown.arrow-up.prevent="onArrowUp"
+    @keydown.enter.prevent="onClick(options[activeFieldIndex])"
+  >
+    <div class="flex pb-3 px-4 border-b-1 border-gray-100">
+      <!-- Todo language for selectFieldToGroup -->
```
[the remark this answers is not in the document]
Comment (rameshmane7218) [reply packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue]: I forgot to remove that after doing corrections, Fixed now

File: packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue (hunk 2 above)
Comment (rameshmane7218) [reply packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue]: done

File: packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue (hunk 3 above)
Comment (rameshmane7218) [reply packages/nc-gui/components/smartsheet/toolbar/CreateGroupBy.vue]: 👍

File: packages/nc-gui/components/smartsheet/toolbar/GroupByMenu.vue (hunk 5 above)
Comment (rameshmane7218) [reply packages/nc-gui/components/smartsheet/toolbar/GroupByMenu.vue]: done

File: packages/nc-gui/helpers/parsers/parserHelpers.ts [hunk 12]
```diff
@@ -191,3 +191,36 @@ export const filterNullOrUndefinedObjectProperties = <T extends Record<string, a
...
+ * @param splitOperator - The separator used to split the defaultName and numbers in existing namesData.
+ *                        Defaults to '-'. Example: If defaultName is 'Token' and splitOperator is '-',
+ *                        existing names like 'Token-1', 'Token-2', etc., will be considered.
+ * @returns The next default name with an incremented number based on existing namesData.
+ */
+export const extractNextDefaultName = (namesData: string[], defaultName: string, splitOperator: string = '-'): string => {
+  // Extract and sort numbers associated with the provided defaultName
+  const extractedSortedNumbers =
+    (namesData
+      .map((name) => {
```
Comment (wingkwong) [inline packages/nc-gui/helpers/parsers/parserHelpers.ts:209]: Array.prototype.map() expects a value to be returned at the end of arrow function

**Review by wingkwong — CHANGES_REQUESTED** [review]
please pull the code once, run lint (`pnpm run lint`) and fix the lint issue(s).

File: packages/nc-gui/lang/en.json [hunk 13]
```diff
@@ -1207,7 +1209,9 @@
       "goToPrevious": "Go to previous",
       "goToNext": "Go to next",
       "thankYou": "Thank you!",
-      "submittedFormData": "You have successfully submitted the form data."
+      "submittedFormData": "You have successfully submitted the form data.",
+      "editingSystemKeynotSupported": "Editing system key not supported"
```
Comment (wingkwong) [inline packages/nc-gui/lang/en.json:1213]: `editingSystemKeyNotSupported`

File: packages/nc-gui/components/cell/DateTimePicker.vue [hunk 14]
```diff
@@ -131,6 +131,10 @@ watch(
   (next) => {
     if (next) {
       onClickOutside(document.querySelector(`.${randomClass}`)! as HTMLDivElement, () => (open.value = false))
+
+      if (!modelValue) {
+        localState.value = dayjs(new Date())
+      }
```
Comment (wingkwong) [inline packages/nc-gui/components/cell/DateTimePicker.vue:137]: as discussed with Raju, the behaviour is a bit off. When I open the datetime picker, the current datetime is expected to be selected. However, we only save the value once we click `OK`, which means the value shouldn't be saved if I un-focus it. Currently the value will be saved anyway.

File: packages/nc-gui/components/smartsheet/column/EditOrAdd.vue [hunk 15]
```diff
@@ -159,6 +159,9 @@ onMounted(() => {
     if (formState.value.pk) {
       message.info(t('msg.info.editingPKnotSupported'))
       emit('cancel')
+    } else if (isSystemColumn(formState.value)) {
+      message.info(t('msg.info.editingSystemKeynotSupported'))
```
Comment (wingkwong) [inline packages/nc-gui/components/smartsheet/column/EditOrAdd.vue:163]: `editingSystemKeyNotSupported`

**Review by wingkwong — CHANGES_REQUESTED** [review]

File: packages/nc-gui/components/cell/DateTimePicker.vue (hunk 14 above)
Comment (rameshmane7218) [reply packages/nc-gui/components/cell/DateTimePicker.vue:137]: Okay, first I tried to set ` :default-picker-value="dayjs(new Date())"` but this is not woking.
I will check this

**Comment by dstala** [comment]
- [ ] Group-by didn't display `Attachment` field earlier. It gets listed now. Please double check on supported fields before and now.

Screenshot 2023-12-10 at 3 51 54 PM

File: packages/nc-gui/components/account/Token.vue [hunk 16]
```diff
@@ -52,6 +62,38 @@ const hideOrShowToken = (tokenId: string) => {
...
+    if (!response) return
+
+    allTokens.value = response.list as IApiTokenInfo[]
+    setDefaultTokenName()
+  } catch (e: any) {
+    message.error(await extractSdkResponseErrorMsg(e))
+  }
+}
+
+// This will updated allTokens local value instead of fetching all tokens on each operation (add|delete)
```
Comment (wingkwong) [inline packages/nc-gui/components/account/Token.vue:82]: `updated` -> `update`

**Review by wingkwong — APPROVED** [review]

File: packages/nc-gui/components/account/Token.vue (hunk 16 above)
Comment (rameshmane7218) [reply packages/nc-gui/components/account/Token.vue]: done

**Review by dstala — APPROVED** [review]

**Pull request merged**